File: Algorithms/HashTable/JewelsAndStones.cpp
//
// Created by archit.s on 13/08/18.
//
// Time:  O(m + n)
// Space: O(n)

#include <cstdlib>
#include <string>
#include <iostream>
#include <unordered_set>

using namespace std;

class JewelsAndStones {
public:
    int numJewelsInStones(string J, string S) {
        unordered_set<char> lookup;
        for (const auto& j : J) {
            lookup.insert(j);
        }
        int result = 0;
        for (const auto& s : S) {
            result += lookup.count(s);
        }
        return result;
    }
};

int main(){

    JewelsAndStones jewelsAndStones;
    cout << jewelsAndStones.numJewelsInStones("aa", "AAaaBB");

    return 0;
}

File: Algorithms/HashTable/JewelsAndStones.java
package HashTable;

import java.util.HashMap;
import java.util.Map;

/**
 * Author - archit.s
 * Date - 14/08/18
 * Time - 11:34 AM
 */
public class JewelsAndStones {
    public int numJewelsInStones(String J, String S) {

        int count = 0;

        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < J.length(); i++) {
            map.put(J.charAt(i), 1);
        }
        for (int i = 0; i < S.length(); i++) {
            Integer integer = map.get(S.charAt(i));
            if(integer != null){
                count += integer;
            }
        }

        return count;
    }

    public static void main(String[] args) {
        System.out.println(new JewelsAndStones().numJewelsInStones("aa", "AAa"));
    }
}
